File: services/fare-pricing/src/fare_pricing/adapters/messaging/subscriber.py
from __future__ import annotations

import json
import logging
import os
import time
from collections import OrderedDict
from collections.abc import Callable
from threading import Event, Thread
from typing import Any

from ...ports import EventEnvelope
from ...ports.messaging import (
    EventSubscriber,
    FatalHandlerError,
    SubscribeFailed,
    TransientHandlerError,
)

logger = logging.getLogger(__name__)


class RedisEventSubscriber(EventSubscriber):
    """Redis Streams implementation of EventSubscriber.

    Uses XREADGROUP consumer groups, XAUTOCLAIM recovery, and DLQ.
    Only Redis types/imports live in this module (contract rule).
    """

    def __init__(self, redis_url: str | None = None, dedup_max_entries: int = 10000) -> None:
        import redis as _redis  # type: ignore[import-untyped]

        self._redis_url = redis_url or os.environ.get("REDIS_URL", "redis://localhost:6379")
        self._client: _redis.Redis | None = None  # type: ignore[name-defined]
        self._stop_event = Event()
        self._threads: list[Thread] = []
        self._dedup_max_entries = dedup_max_entries
        self._seen_event_ids: OrderedDict[str, None] = OrderedDict()

    def _get_client(self) -> Any:  # type: ignore[type-arg]
        import redis as _redis  # type: ignore[import-untyped]

        if self._client is None:
            self._client = _redis.from_url(self._redis_url, decode_responses=True)  # type: ignore[attr-defined]
        return self._client

    def subscribe(
        self,
        streams: list[str],
        group: str,
        consumer_name: str,
        handler: Callable[[EventEnvelope], None],
    ) -> None:
        client = self._get_client()
        import redis as _redis  # type: ignore[import-untyped]

        # Create consumer groups for each stream
        for stream_key in streams:
            try:
                client.xgroup_create(stream_key, group, id="$", mkstream=True)
            except _redis.ResponseError as exc:
                if "BUSYGROUP" not in str(exc):
                    raise SubscribeFailed(f"Failed to create group for {stream_key}: {exc}") from exc
                # BUSYGROUP means the group already exists — safe to ignore

        # Start a polling thread per stream
        for stream_key in streams:
            thread = Thread(
                target=self._poll_stream,
                args=(stream_key, group, consumer_name, handler),
                daemon=True,
            )
            thread.start()
            self._threads.append(thread)

        # Start a recovery thread
        recovery_thread = Thread(
            target=self._recovery_loop,
            args=(streams, group, consumer_name, handler),
            daemon=True,
        )
        recovery_thread.start()
        self._threads.append(recovery_thread)

    def _poll_stream(
        self,
        stream_key: str,
        group: str,
        consumer_name: str,
        handler: Callable[[EventEnvelope], None],
    ) -> None:
        client = self._get_client()
        import redis as _redis  # type: ignore[import-untyped]

        while not self._stop_event.is_set():
            try:
                result = client.xreadgroup(
                    group,
                    consumer_name,
                    {stream_key: ">"},
                    count=10,
                    block=2000,
                )
                if not result:
                    continue

                for stream_name, messages in result:
                    for msg_id, msg_data in messages:
                        if self._stop_event.is_set():
                            return

                        envelope_json = msg_data.get("envelope", "{}")
                        try:
                            envelope_data = json.loads(envelope_json)
                            envelope = EventEnvelope.from_json_dict(envelope_data)
                        except (json.JSONDecodeError, KeyError, ValueError) as exc:
                            logger.warning("Failed to deserialize event from %s: %s", msg_id, exc)
                            # Move unparseable messages to DLQ
                            self._move_to_dlq(client, stream_key, msg_id, envelope_json)
                            client.xack(stream_key, group, msg_id)
                            continue

                        try:
                            if self._already_processed(envelope):
                                client.xack(stream_key, group, msg_id)
                                continue
                            handler(envelope)
                            self._mark_processed(envelope)
                            client.xack(stream_key, group, msg_id)
                        except TransientHandlerError:
                            # Do not XACK; let XAUTOCLAIM retry
                            logger.info("Transient error processing %s, leaving in PEL", msg_id)
                        except FatalHandlerError:
                            # Move to DLQ and XACK
                            self._move_to_dlq(client, stream_key, msg_id, envelope_json)
                            client.xack(stream_key, group, msg_id)
                        except Exception as exc:
                            logger.exception("Unexpected error processing %s: %s", msg_id, exc)
                            # Treat as transient
                            pass
            except _redis.RedisError as exc:
                logger.error("Redis error in polling loop: %s", exc)
                if not self._stop_event.is_set():
                    time.sleep(1)

    def _recovery_loop(
        self,
        streams: list[str],
        group: str,
        consumer_name: str,
        handler: Callable[[EventEnvelope], None],
    ) -> None:
        client = self._get_client()
        import redis as _redis  # type: ignore[import-untyped]

        while not self._stop_event.is_set():
            time.sleep(60)
            if self._stop_event.is_set():
                return

            for stream_key in streams:
                try:
                    claimed = client.xautoclaim(
                        stream_key,
                        group,
                        consumer_name,
                        min_idle_time=60000,
                        start_id="0",
                        count=100,
                    )
                    # xautoclaim returns (next_start_id, [messages...]) in redis-py >= 5
                    if isinstance(claimed, (list, tuple)) and len(claimed) >= 2:
                        messages = claimed[1]
                    else:
                        messages = claimed or []

                    for msg_id, msg_data in messages:
                        delivery_count = self._get_delivery_count(client, stream_key, group, msg_id)
                        envelope_json = msg_data.get("envelope", "{}")

                        if delivery_count >= 5:
                            self._move_to_dlq(client, stream_key, msg_id, envelope_json)
                            client.xack(stream_key, group, msg_id)
                            continue

                        try:
                            envelope_data = json.loads(envelope_json)
                            envelope = EventEnvelope.from_json_dict(envelope_data)
                            if self._already_processed(envelope):
                                client.xack(stream_key, group, msg_id)
                                continue
                            handler(envelope)
                            self._mark_processed(envelope)
                            client.xack(stream_key, group, msg_id)
                        except (json.JSONDecodeError, KeyError, ValueError):
                            self._move_to_dlq(client, stream_key, msg_id, envelope_json)
                            client.xack(stream_key, group, msg_id)
                        except TransientHandlerError:
                            pass
                        except FatalHandlerError:
                            self._move_to_dlq(client, stream_key, msg_id, envelope_json)
                            client.xack(stream_key, group, msg_id)
                        except Exception:
                            pass
                except _redis.RedisError as exc:
                    logger.error("Redis error in recovery loop: %s", exc)

    def _already_processed(self, envelope: EventEnvelope) -> bool:
        if envelope.event_id not in self._seen_event_ids:
            return False
        self._seen_event_ids.move_to_end(envelope.event_id)
        return True

    def _mark_processed(self, envelope: EventEnvelope) -> None:
        self._seen_event_ids[envelope.event_id] = None
        self._seen_event_ids.move_to_end(envelope.event_id)
        while len(self._seen_event_ids) > self._dedup_max_entries:
            self._seen_event_ids.popitem(last=False)

    def _get_delivery_count(
        self,
        client: Any,
        stream_key: str,
        group: str,
        msg_id: str,
    ) -> int:
        """Get delivery count from the PEL entry."""
        import redis as _redis  # type: ignore[import-untyped]

        try:
            pending = client.xpending_range(stream_key, group, min=msg_id, max=msg_id, count=1)
            if pending:
                return pending[0].get("times_delivered", 1)
        except _redis.RedisError:
            pass
        return 1

    def _move_to_dlq(
        self,
        client: Any,
        stream_key: str,
        msg_id: str,
        envelope_json: str,
    ) -> None:
        """Move a poison message to the dead-letter stream."""
        dlq_key = f"{stream_key}:dlq"
        try:
            client.xadd(dlq_key, {"envelope": envelope_json}, maxlen=100000, approximate=True)
        except Exception as exc:
            logger.error("Failed to move message %s to DLQ: %s", msg_id, exc)

    def stop(self) -> None:
        """Graceful shutdown."""
        self._stop_event.set()
        for thread in self._threads:
            thread.join(timeout=5)

File: services/offer-management/src/adapters/messaging/subscriber.ts
// ---------------------------------------------------------------------------
// Redis Streams adapter for EventSubscriber
// Implements the port defined in ../../ports/messaging.ts
// Lives ONLY under adapters/messaging/ — no Redis types outside this module.
// ---------------------------------------------------------------------------

import { type Redis } from "ioredis";

import { SubscribeFailed, type EventEnvelope, type EventHandler, type EventSubscriber } from "../../ports/messaging.js";
import { dlqStreamKey } from "./stream-config.js";

const POLL_TIMEOUT_MS = 2000;
const BATCH_SIZE = 10;
const CLAIM_INTERVAL_MS = 60000;
const MAX_DELIVERY_COUNT = 5;
const MIN_IDLE_MS = 60000;

export class RedisEventSubscriber implements EventSubscriber {
  private redis: Redis;
  private running = false;
  private readonly processedEventIds = new Set<string>();
  private startedPromise: Promise<void>;
  private resolveStarted!: () => void;
  private rejectStarted!: (error: unknown) => void;

  constructor(redis: Redis) {
    this.redis = redis;
    this.startedPromise = new Promise((resolve, reject) => {
      this.resolveStarted = resolve;
      this.rejectStarted = reject;
    });
  }

  started(): Promise<void> {
    return this.startedPromise;
  }

  async subscribe(
    streams: readonly string[],
    group: string,
    consumerName: string,
    handler: EventHandler,
    signal: AbortSignal,
  ): Promise<void> {
    this.running = true;

    // Create consumer groups (ignore BUSYGROUP errors)
    try {
      for (const stream of streams) {
        try {
          await (this.redis as any).xgroup("CREATE", stream, group, "$", "MKSTREAM");
        } catch (err: any) {
          // BUSYGROUP means the group already exists — that's fine
          if (!String(err).includes("BUSYGROUP")) {
            throw new SubscribeFailed(`Failed to create consumer group ${group} on ${stream}`, err);
          }
        }
      }
      this.resolveStarted();
    } catch (error) {
      this.running = false;
      this.rejectStarted(error);
      throw error;
    }

    // Start background recovery loop
    const recoveryTimer = setInterval(async () => {
      if (!this.running) return;
      for (const stream of streams) {
        try {
          await this.claimAndProcess(stream, group, consumerName, handler);
        } catch {
          // log and continue
        }
      }
    }, CLAIM_INTERVAL_MS);

    // Main polling loop
    try {
      while (this.running && !signal.aborted) {
        for (const stream of streams) {
          if (!this.running || signal.aborted) break;
          try {
            await this.pollStream(stream, group, consumerName, handler);
          } catch {
            // log and continue
          }
        }
      }
    } finally {
      clearInterval(recoveryTimer);
    }
  }

  stop(): void {
    this.running = false;
  }

  private async pollStream(
    stream: string,
    group: string,
    consumerName: string,
    handler: EventHandler,
  ): Promise<void> {
    const results: any = await (this.redis as any).xreadgroup(
      "GROUP",
      group,
      consumerName,
      "BLOCK",
      POLL_TIMEOUT_MS,
      "COUNT",
      BATCH_SIZE,
      "STREAMS",
      stream,
      ">",
    );

    if (!results) return;

    for (const [, entries] of results) {
      for (const entry of entries) {
        await this.processEntry(stream, group, entry, handler);
      }
    }
  }

  private async processEntry(
    stream: string,
    group: string,
    entry: [string, string[]],
    handler: EventHandler,
  ): Promise<void> {
    const [entryId, fields] = entry;

    // Find the envelope field
    const envelopeIdx = fields.indexOf("envelope");
    if (envelopeIdx === -1 || envelopeIdx + 1 >= fields.length) {
      // Malformed entry — ack and skip
      await (this.redis as any).xack(stream, group, entryId);
      return;
    }

    let envelope: EventEnvelope;
    try {
      envelope = JSON.parse(fields[envelopeIdx + 1]) as EventEnvelope;
    } catch {
      // Invalid JSON — ack and skip
      await (this.redis as any).xack(stream, group, entryId);
      return;
    }

    if (this.processedEventIds.has(envelope.eventId)) {
      await (this.redis as any).xack(stream, group, entryId);
      return;
    }

    const result = await handler(envelope);

    switch (result) {
      case "ack":
        this.processedEventIds.add(envelope.eventId);
        await (this.redis as any).xack(stream, group, entryId);
        break;
      case "dlq": {
        await this.moveToDlq(stream, group, [entryId, fields], JSON.stringify(envelope));
        break;
      }
      case "retry":
        // Leave in PEL for retry via XAUTOCLAIM
        break;
    }
  }

  private async claimAndProcess(
    stream: string,
    group: string,
    consumerName: string,
    handler: EventHandler,
  ): Promise<void> {
    const result: any = await (this.redis as any).xautoclaim(
      stream,
      group,
      consumerName,
      MIN_IDLE_MS,
      "0",
      "COUNT",
      100,
    );

    const entries = result[1] as Array<[string, string[]]>;
    if (!entries || entries.length === 0) return;

    const deliveryCounts = await this.deliveryCounts(stream, group, entries.map(([entryId]) => entryId));
    for (const [entryId, fields] of entries) {
      if ((deliveryCounts.get(entryId) ?? 1) >= MAX_DELIVERY_COUNT) {
        await this.moveToDlq(stream, group, [entryId, fields]);
        continue;
      }

      await this.processEntry(stream, group, [entryId, fields], handler);
    }
  }

  private async deliveryCounts(stream: string, group: string, entryIds: readonly string[]): Promise<Map<string, number>> {
    if (entryIds.length === 0) {
      return new Map();
    }

    const counts = new Map<string, number>();
    await Promise.all(entryIds.map(async (entryId) => {
      const pending: any = await (this.redis as any).xpending(stream, group, entryId, entryId, 1);
      const row = Array.isArray(pending) ? pending[0] : undefined;
      if (Array.isArray(row) && row[0] === entryId) {
        const deliveries = Number(row[3]);
        if (Number.isFinite(deliveries)) {
          counts.set(entryId, deliveries);
        }
      }
    }));
    return counts;
  }

  private async moveToDlq(
    stream: string,
    group: string,
    entry: [string, string[]],
    envelopeJson?: string,
  ): Promise<void> {
    const [entryId, fields] = entry;
    const envelopeIdx = fields.indexOf("envelope");
    const serializedEnvelope = envelopeJson ?? (envelopeIdx !== -1 ? fields[envelopeIdx + 1] : undefined);

    if (serializedEnvelope) {
      await (this.redis as any).xadd(
        dlqStreamKey(producerFromStream(stream)),
        "MAXLEN",
        "~",
        "100000",
        "*",
        "envelope",
        serializedEnvelope,
      );
    }

    await (this.redis as any).xack(stream, group, entryId);
  }

}

function producerFromStream(stream: string): string {
  return stream.startsWith("events:") ? stream.slice("events:".length).replace(/:dlq$/, "") : stream;
}
